discord/voice_client.py

```diff
@@ -291,8 +306,188 @@ def is_connected(self):
+        return coro
+
     # audio related
 
+    @asyncio.coroutine
+    def get_voice_data(self, user, timeout=400):
+        """|coro|
+
+        Receive voice data from a user untill they are silent for
+        `timeout` ms.
+
+        Parameters
+        -----------
+        user: :class:`abc.User`
+            The user to listen for audio from.
+        timeout: Optional[int]
+            The time in ms of silence after which to stop listening.
+
+        Returns
+        -----------
+        The received voice data for the user passed in `bytes` form.
+
+        The data is stereo 48kHz audio in raw PCM. You will have to add your
+        own headers if you want to save it as a `.wav` file.
+        """
+        global buffer, last_packet
+        data = b''
+        buffer = []
+        current_timestamp = 0
+        last_packet = 0
+
+        def listener(sequence, timestamp, s_user, decoded_data):
+            global buffer, last_packet
+            if s_user == user.id:
+                last_packet = time.time() * 1000
+
+                heapq.heappush(buffer, (timestamp, decoded_data))
+
+        self._pcm_listeners.append(listener)
+
+        while (time.time() * 1000) - last_packet < timeout or last_packet == 0:
+            yield from asyncio.sleep(0.1)  # Continue to fill up the buffer.
+        self._pcm_listeners.remove(listener)
+
+        buffer.sort()
+        for key, value in buffer:
+            # if current_timestamp == 0:
+            #     current_timestamp = key
+            #     delta = 0
+            # else:
+            #     delta = max(0, key - current_timestamp - 960)
+            # padding = bytearray([0] * delta)
+            # data += padding
+            data += value
+
+        return data
+
+    def _decode_voice_packet(self, data):
+        """
+        Decodes a voice packet returning the header attributes and the
+        voice data (pre-opus) after going through NaCl.
+        """
+
+        header = data[:12]
+        data = data[12:]
+        nonce = bytearray(24)
+        box = nacl.secret.SecretBox(bytes(self.secret_key))
+
+        # Decode header
+        type_ = header[0]
+        version = header[1]
+        sequence = struct.unpack(">H", header[2:4])[0]
+        timestamp = struct.unpack(">I", header[4:8])[0]
+        SSRC = struct.unpack(">I", header[8:12])[0]
+
+        # Copy header to nonce's first 12 bytes
+        nonce[:12] = header
+
+        decrypted_data = box.decrypt(bytes(data), bytes(nonce))
+
+        # Encrypt and return the data
+        return type_, version, sequence, timestamp, SSRC, decrypted_data
+
+    def _voice_receive_loop(self):
+        """
+        Listener loop.
+        This loop will poll `self.socket` until there is some data,
+        then it will unpack the data, decode it, and send it to the
+        user-registered callback(s).
+        """
+
+        while self.socket is None:
+            time.sleep(0.1)
```

Comment:
Consider using a `threading.Event` instead of a busy short sleep loop.

See my comment above about refactoring this design so it follows the one from `AudioPlayer`.